lib/hl7/datetime_parser.rb

```diff
@@ -0,0 +1,62 @@
+require "hl7/datetime_components"
+require "active_support/core_ext/module/delegation"
+
+module HL7
+  class DatetimeParser
+    # The HL7 datetime format is:
+    #    YYYY[MM[DD[HH[MM[SS[.S[S[S[S]]]]]]]]][+/-ZZZZ]
+    RX = /\A(\d{4})(?:(\d{2})(?:(\d{2})(?:(\d{2})(?:(\d{2})(?:(\d{2})(\.\d{1,4})?)?)?)?)?)?([-+]\d{4})?\Z/.freeze
+
+    def initialize(text)
+      @text = text
+    end
+
+    def components
+      @components ||= (
+        y, mo, d, h, mi, s, f, off = match
+
+        DatetimeComponents.new(
+          year: y.to_i,
+          month: (mo.nil? ? 1 : mo.to_i),
+          day: (d.nil? ? 1 : d.to_i),
+          hour: h.to_i,
+          minute: mi.to_i,
+          second: (s.to_i + f.to_f),
+          offset: offset_seconds(off)
+        )
+      )
+    end
+
+    delegate :to_time, to: :components
+
+    private
+
+    attr_reader :text
+
+    def match
+      # Note that: nil.to_a   => []
+      #       and: [].drop(1) => []
+      text.match(RX).to_a.drop(1)
```

Comment:
Might be worth adding a note that the returned `MatchData` has the original string as it's first element, hence the drop. I had forgotten that and had to look it up.

Reply:
Yep, good point.